src/format.cc

```diff
@@ -68,74 +89,166 @@ uint32_t Format::InputNeededPerElement() const {
   return count;
 }
 
-void Format::AddComponent(FormatComponentType type,
-                          FormatMode mode,
-                          uint8_t bits) {
-  components_.push_back(MakeUnique<Component>(type, mode, bits));
-  RebuildSegments();
-}
-
-void Format::SetColumnCount(uint32_t c) {
-  column_count_ = c;
-  RebuildSegments();
-}
-
 void Format::SetLayout(Layout layout) {
   layout_ = layout;
   RebuildSegments();
 }
 
 void Format::RebuildSegments() {
   segments_.clear();
+  AddSegmentForType(type_);
+}
+
+void Format::AddPaddedSegment(uint32_t size) {
+  segments_.push_back(Segment{size});
+}
+
+bool Format::NeedsPadding(type::Type* t) const {
+  if (layout_ == Layout::kStd140 && (t->IsMatrix() || t->IsArray()))
+    return true;
+  if (t->IsVec3() || (t->IsMatrix() && t->RowCount() == 3))
+    return true;
+  return false;
+}
 
-  for (size_t i = 0; i < column_count_; ++i) {
-    for (size_t k = 0; k < components_.size(); ++k) {
-      segments_.push_back(Segment{components_[k].get()});
+uint32_t Format::AddSegmentForType(type::Type* type) {
+  if (type->IsList() && type->AsList()->IsPacked()) {
+    auto l = type->AsList();
+    segments_.push_back(Segment(FormatComponentType::kR, FormatMode::kUInt,
+                                l->PackSizeInBits()));
+    return l->SizeInBytes();
+  }
+
+  if (type->IsStruct()) {
+    auto s = type->AsStruct();
+    uint32_t cur_offset = 0;
+    for (const auto& member : s->Members()) {
+      if (member.HasOffset() &&
+          static_cast<uint32_t>(member.offset_in_bytes) > cur_offset) {
+        AddPaddedSegment(static_cast<uint32_t>(member.offset_in_bytes) -
+                         cur_offset);
+        cur_offset = static_cast<uint32_t>(member.offset_in_bytes);
+      }
```

Comment:
It looks like you're assuming lower members have lower offsets. That's completely reasonable, but not required of the SPIR-V decorations. You should probably add an assert that the member offset is >= cur_offset.

Reply:
Specifically Vulkan 14.5.4 "Offset and Stride Assignment"

> The numeric order of `Offset` decorations does not need to follow member declaration order.

Reply:
Done, line 260